File: service/src/main/java/coding/challenge/core/context/ExecutionContext.java
package coding.challenge.core.context;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.Locale;

/**
 * Contains information about execution context fo the current request.
 */
@Setter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public final class ExecutionContext {

    private String language;

    private static final ThreadLocal<ExecutionContext> CONTEXT = ThreadLocal.withInitial(ExecutionContext::new);

    public static ExecutionContext get() {
        return CONTEXT.get();
    }

    public static void clear() {
        CONTEXT.remove();
    }

    /**
     * Returns language code specified in the request or default locale language.
     *
     * @return Language code
     */
    public String getLanguage() {
        return language == null ? Locale.getDefault().getLanguage() : language;
    }

}

File: service-kt/src/main/kotlin/coding/challenge/core/context/ExecutionContext.kt
package coding.challenge.core.context

import java.util.Locale

/**
 * Contains information about execution context fo the current request.
 */
class ExecutionContext {

    /**
     * Language code specified in the request or default locale language.
     */
    var language: String? = null
        get() = field ?: Locale.getDefault().language

    companion object {

        private val CONTEXT: ThreadLocal<ExecutionContext> = ThreadLocal.withInitial { ExecutionContext() }

        fun get(): ExecutionContext {
            return CONTEXT.get()
        }

        fun clear() {
            CONTEXT.remove()
        }

    }

}
